source/common/config/utility.cc

```diff
@@ -81,13 +81,32 @@ void Utility::checkFilesystemSubscriptionBackingPath(const std::string& path) {
 void Utility::checkApiConfigSourceSubscriptionBackingCluster(
     const Upstream::ClusterManager::ClusterInfoMap& clusters,
     const envoy::api::v2::core::ApiConfigSource& api_config_source) {
-  if (api_config_source.cluster_names().size() != 1) {
-    // TODO(htuch): Add support for multiple clusters, #1170.
-    throw EnvoyException(
-        "envoy::api::v2::core::ConfigSource must have a singleton cluster name specified");
+  bool is_grpc = (api_config_source.api_type() == envoy::api::v2::core::ApiConfigSource::GRPC);
+
+  if (is_grpc) {
+    if (api_config_source.cluster_names().size() != 0) {
+      throw EnvoyException(
+          "envoy::api::v2::core::ConfigSource::GRPC must not have a cluster name specified");
+    }
+    if (api_config_source.grpc_services().size() != 1) {
+      throw EnvoyException(
+          "envoy::api::v2::core::ConfigSource::GRPC must have a singleton grpc service specified");
+    }
+  } else {
+    if (api_config_source.grpc_services().size() != 0) {
+      throw EnvoyException(
+          "envoy::api::v2::core::ConfigSource must not have a grpc service specified");
+    }
+    if (api_config_source.cluster_names().size() != 1) {
+      // TODO(htuch): Add support for multiple clusters, #1170.
+      throw EnvoyException(
+          "envoy::api::v2::core::ConfigSource must have a singleton cluster name specified");
+    }
   }
 
-  const auto& cluster_name = api_config_source.cluster_names()[0];
+  const std::string& cluster_name =
+      is_grpc ? api_config_source.grpc_services()[0].envoy_grpc().cluster_name()
```

Comment:
There needs to be another condition here on `has_envoy_grpc()`. Some gRPC services will be configured with `google_grpc` and won't have a cluster name.